server/src/main/java/org/elasticsearch/cluster/service/ClusterApplierRecordingService.java

```diff
@@ -75,14 +87,39 @@ Releasable record(String action) {
 
             this.recording = true;
             this.currentAction = action;
-            this.startTimeMS = currentTimeSupplier.getAsLong();
+            this.startMillis = threadPool.rawRelativeTimeInMillis();
+
+            if (logger.isDebugEnabled()) {
+                currentListener = new SubscribableListener<>();
+                currentListener.addTimeout(debugLoggingTimeout, threadPool, threadPool.generic());
+                currentListener.addListener(new ActionListener<>() {
+                    @Override
+                    public void onResponse(Void unused) {}
+
+                    @Override
+                    public void onFailure(Exception e) {
```

Comment:
nit: an assert that it is a timeout exception would also help understand the code better.

Reply:
++ done in f9bd56f3dc55c3b7d562fecd1bf3e29b8850c0f6.